### PR TITLE
refactor: update fa3 codebase and fix hopper unittest [part 1]

### PR DESCRIPTION
**Pull request by yzh119**
## 📌 Description

This PR refactors the out-dated fa3 codebase, more specifically, for page_size>1, the page offset calculation is performed inside the kernel, without the need of a standalone function call to block_sparse_indices_to_vector_sparse_offsets, and optimize the kv_offset calculation with prefetching and shuffling.

This PR also fixes the failed unittest on hopper.

However, the FA3 structure in our codebase is still terrible outdated without important features such as `IntraWGOverlap` and `RescaleOBeforeGemm`, will follow up soon in a later PR.

## 🔍 Related Issues

This PR should fixes https://github.com/flashinfer-ai/flashinfer/issues/1647

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

## Summary by CodeRabbit

* **Bug Fixes**
  * Enforced consistent K/V page-strides and stride_n for sparse paged attention with runtime validation.

* **New Features**
  * Exposed page-interval stride metadata for paged KV caches.
  * Expanded FP8 support: per-head/per-tensor scale propagation and configurable output dtype; added ragged-KV execution path and benchmarks.

* **Refactor**
  * Unified sparse attention to a page-based addressing model and simplified workspace/epilogue behavior; removed legacy block-to-vector conversion helpers and related APIs.

* **Tests**
  * Added FP8 ragged/paged tests, expanded page-size coverage, and removed obsolete vector-sparse unit test.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Review by yongwww — APPROVED** [review]

File: include/flashinfer/attention/hopper/epilogue.cuh [hunk 1]
```diff
@@ -194,11 +194,10 @@ struct CollectiveEpilogue {
       }
     }
 
+    cutlass::arch::NamedBarrier::sync(NUM_MMA_THREADS,
+                                      cutlass::arch::ReservedNamedBarriers::EpilogueBarrier);
+
     int write_warp_idx = NUM_WARPS - 1;
```
Comment (yongwww) [inline include/flashinfer/attention/hopper/epilogue.cuh]: nit: looks like `write_warp_idx` is unused.

Comment (yzh119) [reply include/flashinfer/attention/hopper/epilogue.cuh]: No it was used in https://github.com/flashinfer-ai/flashinfer/pull/2111/files#diff-1577aba07c747a85f70968de2954372f63adba262f77a6f563cbbc07cd456ab7R204

File: csrc/batch_prefill_fp8_ragged_sm90_kernel_inst.jinja [hunk 2]
```diff
@@ -1 +1,15 @@
...
+{% for same_scheduler_for_all_heads in ["true", "false"] %}
+template cudaError_t BatchFP8PrefillWithRaggedKVCacheDispatched
+    <{{ head_dim_qk }},
+     {{ mask_mode }},
+     /*USE_SLIDING_WINDOW=*/{{ use_sliding_window }},
+     /*SAME_SCHEDULER_FOR_ALL_HEADS=*/{{ same_scheduler_for_all_heads }},
+     {{ variant_name }}, RaggedParams>(RaggedParams& params, bool enable_pdl, cudaStream_t stream);
+{% endfor %}
+
+};  // namespace flashinfer
```
[the remark this answers is not in the document]
Comment (nvpohanh) [reply csrc/batch_prefill_fp8_ragged_sm90_kernel_inst.jinja]: coderabbit is correct

File: flashinfer/prefill.py [hunk 3]
```diff
@@ -1627,6 +1648,9 @@ def plan(
             The data type of the query tensor, defaults torch.float16.
         kv_data_type : Optional[Union[str, torch.dtype]]
             The data type of the key/value tensor. If None, will be set to :attr:`q_data_type`.
+        o_data_type : Optional[Union[str, torch.dtype]]
+            The data type of the output tensor. If None, will be set to :attr:`q_data_type`.
+            For FP8 inputs, this should typically be set to torch.float16.
```
Comment (nvpohanh) [inline flashinfer/prefill.py:1704]: `torch.float16` -> `torch.float16 or torch.bfloat16`

File: flashinfer/prefill.py [hunk 4]
```diff
@@ -1543,6 +1563,7 @@ def plan(
         rope_theta: Optional[float] = None,
         q_data_type: Union[str, torch.dtype] = "float16",
         kv_data_type: Optional[Union[str, torch.dtype]] = None,
+        o_data_type: Optional[Union[str, torch.dtype]] = None,
```
Comment (nvpohanh) [inline flashinfer/prefill.py:1617]: In addition to prefill.py, we should also update decode.py. See https://github.com/flashinfer-ai/flashinfer/pull/2148 as an example. thanks

File: flashinfer/prefill.py [hunk 5]
```diff
@@ -2042,13 +2054,10 @@ def run(
             q, k_cache, self._cached_q_data_type, self._cached_kv_data_type
```
Comment (nvpohanh) [inline flashinfer/prefill.py:2106]: We should add docs for `q_scale`

File: flashinfer/prefill.py [hunk 6]
```diff
@@ -2173,6 +2167,14 @@ def run(
             if self._jit_module is not None:
                 run_args.extend(list(args))
             else:
+                # Extract FP8 scale tensors from *args if q is FP8
+                fp8_scale_q = None
+                fp8_scale_k = None
+                fp8_scale_v = None
+                if is_float8(q) and len(args) >= 3:
```
Comment (nvpohanh) [inline flashinfer/prefill.py:2226]: I am not a fan of making these are implicit positional args. Instead, I think we should document it clearly and make them explicit.

For example, we can make `k_scale : Optional[torch.Tensor, float]` or have separate keyword args like `k_scale_device`. Also, we should document clearly that these scale tensors should be per-head. Otherwise, illegal mem access will occur.

File: flashinfer/prefill.py [hunk 7]
```diff
@@ -2198,7 +2200,7 @@ def run(
                     self._max_kv_len,
                     self._batch_size,
                     self._qo_indptr_buf,
-                    self._vector_sparse_indptr_buffer,
+                    self._paged_kv_indptr_buf,
                     sinks,
                 ]
```
Comment (nvpohanh) [inline flashinfer/prefill.py:2258]: at line 2209, small yet important optimization: `if v_scale is not None:` -> `if v_scale is not None and v_scale != 1.0:`

File: flashinfer/prefill.py [hunk 8]
```diff
@@ -2587,6 +2590,9 @@ def plan(
             The data type of the query tensor, defaults to torch.float16.
         kv_data_type : Optional[Union[str, torch.dtype]]
             The data type of the key/value tensor. If None, will be set to :attr:`q_data_type`.
+        o_data_type : Optional[Union[str, torch.dtype]]
+            The data type of the output tensor. If None, will be set to :attr:`q_data_type`.
+            For FP8 inputs, this should typically be set to torch.float16.
```
Comment (nvpohanh) [inline flashinfer/prefill.py:2649]: `torch.float16` -> `torch.float16 or torch.bfloat16`

File: include/flashinfer/attention/hopper/quantization/mainloop_sparse_load.cuh [hunk 9]
```diff
@@ -208,44 +208,122 @@ struct FP8SparseCollectiveMainloop {
...
-        make_shape(kv_len, HEAD_DIM), stride<0>(mainloop_params.layout_K), indexed_gather);
-    Tensor mV = make_block_sparse_tensor(  // (kv_len, D)
-        make_gmem_ptr(mainloop_params.V_ptr + kv_head_idx * stride<2>(mainloop_params.layout_V)),
-        make_shape(kv_len, HEAD_DIM), stride<0>(mainloop_params.layout_V), indexed_gather);
-
-    Tensor gK = local_tile(mK, select<1, 2>(TileShape_QKD{}), make_coord(_, _0{}));  // (KV, D, kv)
-    Tensor gV = local_tile(mV, select<1, 2>(TileShape_QKD{}), make_coord(_, _0{}));  // (KV, D, kv)
-    Tensor cKV = cute::make_identity_tensor(gK.shape());
+    // Setup for manual K/V loading with page table
+    DTypeKV* k_base_ptr = mainloop_params.K_ptr;
```
Comment (nvpohanh) [inline include/flashinfer/attention/hopper/quantization/mainloop_sparse_load.cuh]:
```
    DTypeKV* k_base_ptr = mainloop_params.K_ptr + kv_head_idx * HEAD_DIM;
    DTypeKV* v_base_ptr = mainloop_params.V_ptr + kv_head_idx * HEAD_DIM;
```

This is an accuracy bug. The base ptr should add `kv_head_idx * HEAD_DIM`

File: flashinfer/prefill.py [hunk 10]
```diff
@@ -2182,9 +2184,9 @@ def run(
                     self._max_item_len_ptr,
                     logits_soft_cap,
                     sm_scale,
-                    None,  # scale_q, not supported yet
-                    None,  # scale_k
-                    None,  # scale_v
+                    fp8_scale_q,
```
Comment (nvpohanh) [inline flashinfer/prefill.py:2239]: In the FA3 FP8 kernels, I suggest that we change `scale_q`/`scale_k`/`scale_v` to `maybe_scale_q`/`maybe_scale_k`/`maybe_scale_v`

See https://github.com/flashinfer-ai/flashinfer/pull/2148/commits/a8d9e6a8f6378a95ed5526bd2403c8c00f0134fe as an example about how to do this.

File: tests/attention/test_hopper_fp8_attention.py [hunk 11]
```diff
@@ -183,7 +183,248 @@ def test_block_sparse_attention(
...
+        pytest.skip("SM90A is not supported")
+
+    print(
+        f"Testing FP8 batch prefill ragged with batch_size={batch_size}, num_heads={num_heads}, "
+        f"head_dim={head_dim}, causal={causal}, dtype={dtype}"
+    )
+
+    # Setup
+    o_dtype = torch.half
+    num_qo_heads = num_kv_heads = num_heads
```
Comment (nvpohanh) [inline tests/attention/test_hopper_fp8_attention.py:267]: we should at least have a test case with GQA setup (i.e. `num_qo_heads != num_kv_heads` and `num_qo_heads % num_kv_heads == 0`) for better coverage

File: tests/attention/test_hopper_fp8_attention.py [hunk 12]
```diff
@@ -183,7 +183,248 @@ def test_block_sparse_attention(
...
+        num_qo_heads,
+        num_kv_heads,
+        head_dim,
+        head_dim,
+        q_data_type=dtype,
+        kv_data_type=dtype,
+        o_data_type=o_dtype,
+        causal=causal,
+    )
+    o_fp8 = wrapper_fp8.run(q_fp8, k_fp8, v_fp8, s_q, s_k, s_v)
```
Comment (nvpohanh) [inline tests/attention/test_hopper_fp8_attention.py:340]: We should have tests that cover both types of scaling factors:

1. host-side per-tensor scales
2. device-side per-head scale tensors

File: include/flashinfer/attention/hopper/quantization/mainloop_sparse_load.cuh (hunk 9 above)
Comment (yzh119) [reply include/flashinfer/attention/hopper/quantization/mainloop_sparse_load.cuh]: Fixed in https://github.com/flashinfer-ai/flashinfer/pull/2111/commits/594a4677e2b16d28c92ff3bb23efb835ba10032f (and improve unittests), actually it should be `kv_head_idx  * stride` instead of `HEAD_DIM`, the stride is not necessarily HEAD_DIM, because we support different kv cache layouts.

File: tests/attention/test_hopper_fp8_attention.py (hunk 11 above)
Comment (yzh119) [reply tests/attention/test_hopper_fp8_attention.py:267]: done

File: flashinfer/prefill.py (hunk 4 above)
Comment (yzh119) [reply flashinfer/prefill.py:1617]: that's out of the scope of this PR itself.

Comment (nvpohanh) [reply flashinfer/prefill.py:1617]: got it. I will do that in my PR instead. thanks
